packages/sources/superstate/src/transport/nav.ts

```diff
@@ -0,0 +1,201 @@
+    this.requester = dependencies.requester
+    this.settings = settings
+    this.endpointName = endpointName
+    this.fundIdsSet = new Set()
+    this.runScheduler()
+  }
+
+  // registerRequest is invoked on every valid request to EA
+  // Adds fundId in the request to a SET
+  async registerRequest(req: AdapterRequest<typeof inputParameters.validated>) {
+    const { fundId } = req.requestContext.data
+    if (!this.fundIdsSet.has(fundId)) {
+      this.fundIdsSet.add(fundId)
+      logger.info(`Added new fund id - ${fundId}`)
+    }
+  }
+
+  // foregroundExecute is executed when there is a new request/fundId that is not in the cache
+  async foregroundExecute(
+    req: AdapterRequest<typeof inputParameters.validated>,
+  ): Promise<AdapterResponse<BaseEndpointTypes['Response']> | void> {
+    const { fundId } = req.requestContext.data
+    return this.execute(fundId)
+  }
+
+  // Runs 'execute' function every day at 9:09 AM ET (if fundIdsSet is not empty)
+  runScheduler() {
+    const rule = new schedule.RecurrenceRule()
+    rule.hour = 9
+    rule.minute = 9
+    rule.tz = TZ
+
+    schedule.scheduleJob(rule, () => {
+      logger.info(
+        `Scheduled execution started at ${Date.now()}. FundIdSet - ${[...this.fundIdsSet].join(
+          ',',
+        )}`,
+      )
+      ;[...this.fundIdsSet].map(async (fundId) => this.execute(fundId))
+    })
+  }
+
+  // execute is either called by scheduler or foregroundExecute.
+  // Makes a request to DP and saves the response in the cache.
+  // In case the DP returns stale data the function will be executed again several times
+  // before finalizing and saving the last returned data to a cache.
+  async execute(fundId: number, retryCount = 0) {
+    const providerDataRequestedUnixMs = Date.now()
+    const apiResponse = await this.makeRequest(fundId)
+    const providerDataReceivedUnixMs = Date.now()
+
+    if (!apiResponse.data?.length) {
+      const response = {
+        errorMessage: `The data provider did not return any value for fundId: ${fundId}`,
+        statusCode: 502,
+        timestamps: {
+          providerDataRequestedUnixMs,
+          providerDataReceivedUnixMs,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+      await this.responseCache.write(this.name, [{ params: { fundId }, response }])
+      return
+    }
+
+    const data = apiResponse.data[0]
+    const result = Number(data.net_asset_value)
+
+    // DP updates previous working day's price on the next working day at 9:09 AM ET
+    // If there is no fresh price data, we try to re-fetch the API until 10:30 AM ET
+    // Skips checks and re-running on weekends
+    if (this.shouldCheckForExpectedData()) {
+      // At this point we know that conditions are met, and we should check for stale data and retry if it's not updated.
+      // If the most recent update from DP is not from previous working day, the data might be stale
+      const expectedDate = getPreviousNonWeekendDay(TZ)
```

Comment:
Are there any non-Weekend days that might not produce a report and break this assumption, e.g. holiday days?

Reply:
Holidays are not counted here, we will treat them as business days. This means that on holidays and on the next business days we will fetch the DP, mark the response as stale and make retries until 10:30 after which save the last data from DP (which was already cached for previous day) in the cache.